core/src/main/java/org/apache/iceberg/rest/HTTPClient.java

```diff
@@ -0,0 +1,282 @@
+import org.apache.hc.core5.http.io.entity.StringEntity;
+import org.apache.hc.core5.io.CloseMode;
+import org.apache.iceberg.exceptions.RESTException;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.rest.responses.ErrorResponse;
+import org.apache.iceberg.rest.responses.ErrorResponseParser;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An HttpClient for usage with the REST catalog.
+ */
+public class HTTPClient implements RESTClient {
+
+  private static final Logger LOG = LoggerFactory.getLogger(HTTPClient.class);
+
+  private final String uri;
+  private final CloseableHttpClient httpClient;
+  private final ObjectMapper mapper;
+  private final Map<String, String> additionalHeaders;
+
+  private HTTPClient(
+      String uri, CloseableHttpClient httpClient, ObjectMapper mapper, Map<String, String> additionalHeaders) {
+    this.uri = uri;
+    this.httpClient = httpClient != null ? httpClient : HttpClients.createDefault();
+    this.mapper = mapper != null ? mapper : new ObjectMapper();
+    this.additionalHeaders = additionalHeaders != null ? additionalHeaders : ImmutableMap.of();
+  }
+
+  private static String extractResponseBodyAsString(CloseableHttpResponse response) {
+    try {
+      if (response.getEntity() == null) {
+        return null;
+      }
+
+      // EntityUtils.toString returns null when HttpEntity.getContent returns null.
+      return EntityUtils.toString(response.getEntity(), "UTF-8");
+    } catch (IOException | ParseException e) {
+      throw new RESTException(e, "Failed to convert HTTP response body to string");
+    }
+  }
+
+  // Per the spec, the only currently defined / used "success" responses are 200 and 202.
+  private static boolean isSuccessful(CloseableHttpResponse response) {
+    int code = response.getCode();
+    return code ==  HttpStatus.SC_OK || code == HttpStatus.SC_ACCEPTED;
+  }
+
+  private static ErrorResponse buildDefaultErrorResponse(CloseableHttpResponse response) {
+    String responseReason = response.getReasonPhrase();
+    String message =
+        responseReason != null && !responseReason.isEmpty() ? responseReason :
+            EnglishReasonPhraseCatalog.INSTANCE.getReason(response.getCode(), null /* ignored */);
+    String type = "RESTException";
+    return ErrorResponse.builder()
+        .responseCode(response.getCode())
+        .withMessage(message)
+        .withType(type)
+        .build();
+  }
+
+  private static void handleFailure(
+      CloseableHttpResponse response, String responseBody, Consumer<ErrorResponse> errorHandler) {
+    ErrorResponse errorResponse = null;
+
+    if (responseBody != null) {
+      try {
+        errorResponse = ErrorResponseParser.fromJson(responseBody);
+      } catch (Exception e) {
```

Comment:
What exceptions are expected here? Could this catch those exceptions instead?

Reply:
Below, the object mapper calls are wrapped to catch `JsonProcessingException`.

Reply:
This throws `JsonProcessingException` and `JsonMappingException` in the call to reading the `resposneBody` as `JsonNode`, but it catches those (as they're subclasses of `IOException`) and rethrows them as `UncheckedIOExcepion`.

Looking more closely, it can also throw `IllegalArgumentException` due to some precondition checks.

So I think if we catch `UncheckedIOException` and `IllegalArgumentException`, that should cover everything as it currently is.

I'm also thinking of updating `ErrorResponseParser` to catch those `IllegalArgumentException` and rethrow them as `UncheckedIOException` to simplify the possible exceptions it can throw.

Reply:
Updated to catch both `IllegalArgumentException` and `UncheckedIOException`.

I'm making myself a note to add more tests for this in a separate PR to ensure that this is all that can be thrown, but it sees like it.